enhancements/network/user-defined-network-segmentation.md

```diff
@@ -379,11 +382,264 @@ any user-defined primary network.
+
+##### Status
+The CRD status should reflect the NAD state through conditions.
+For example, when the NAD its created the condition should be placed with status `True`.
+
+For cluster scoped networks, the condition should be true once all desired namespaces are provisioned with the corresponding NAD.
+
+The cluster scoped CRD status should reflect on which namespaces the network is available.
+
+The CRD status should reflect the user-defined-network state through conditions, 
+reflecting the NAD state and underlying OVN network state.
+
+##### CRD Definitions
+
+The following CRDs should be installed by [CNO](https://github.com/openshift/cluster-network-operator) and managed by OVN-Kubernetes.
+
+##### Namespace scoped CRD
+```golang
+// UserDefinedNetwork describe network request for a Namespace
+type UserDefinedNetwork struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+    // +kubebuilder:validation:Required
+    // +kubebuilder:validation:XValidation:rule="self == oldSelf", message="Spec is immutable"
+    // +kubebuilder:validation:XValidation:rule=" self.topology != 'Layer3' || (self.topology == 'Layer3' && size(self.subnets) > 0)", message="Subnets is required for Layer3 topology"
+    // +kubebuilder:validation:XValidation:rule="self.topology != 'Localnet' || (self.topology == 'Localnet' && self.role == 'Secondary')", message="Topology localnet is not supported for primary network"
+    // +kubebuilder:validation:XValidation:rule="!has(self.ipamLifecycle) || (self.ipamLifecycle == 'Persistent' && (self.topology == 'Layer2' || self.topology == 'Localnet'))", message="ipamLifecycle is supported for Layer2 and Localnet topologies"
+    // +required
+    Spec UserDefinedNetworkSpec `json:"spec"`
+    // +optional
+    Status UserDefinedNetworkStatus `json:"status,omitempty"`
+}
+```
+Suggested API validation rules:
+- Spec defined as immutable
+  Avoid incomplete state in a scenario where a NAD is created according to UDN spec, and pods connected to the network, 
+  UDN spec changes, now pods connected to a network that was created from previous revision spec.
+- `Subents` are mandatory for `Layer3` topology.
+- `Localnet` topology is not supported for primary network.
+- `IPAMLifecycle` is supported for `Layer2` and `Localnet` topology.
+
+```golang
+// UserDefinedNetworkSpec defines the desired state of UserDefinedNetwork.
+type UserDefinedNetworkSpec struct {
+	// The topological configuration for the network.
+	// +kubebuilder:validation:Enum=Layer2;Layer3;Localnet
+	Topology NetworkTopology `json:"topology"`
+
+	// The network role in the pod (e.g.: Primary, Secondary).
+	// +kubebuilder:validation:Enum=Primary;Secondary
+	Role NetworkRole `json:"role"`
+
+	// The maximum transmission unit (MTU).
+	// The default value is 1400.
+	// +optional
+	Mtu uint `json:"mtu,omitempty"`
```

Comment:
Do we need max value validation for MTU?

Reply:
I wasn't sure about it, what should be the max value?

Reply:
Our min and max values for MTU as specified in CNO are:
const (
	MinMTUIPv4 uint32 = 576  // RFC 791
	MinMTUIPv6 uint32 = 1280 // RFC 8200
	MaxMTU     uint32 = 65536
)

Reply:
Also, the definition of the MTU value in OVNK with CNO:
	// mtu is the MTU to use for the tunnel interface. This must be 100
	// bytes smaller than the uplink mtu.
	// Default is 1400
	// +kubebuilder:validation:Minimum=0
	// +optional
	MTU *uint32 `json:"mtu,omitempty"`

Reply:
Thanks for the refs

We good on not having validation on max value because CNO dont enforce it?
Or we should add `MTU <= 65536` validation?

Reply:
OK I think I've added more confusion with my comments here. Let me try to to clarify. 

So in CNO:
1. We check if MTU is within the boundaries of those constants.
2. We do not default to 1400 (even though the comment says that). We use this mtu-prober program that will probe a host in OCP to figure out what the MTU should be.

In OVNK:
1. We default the MTU globally as 1400 if it is not provided to us.
2. We have no checks to make sure the MTU is within a min/max boundary. Perhaps we should change that though.

Now, this UDN CRD is an OVNK API and not an OCP API. Therefore we should not reference OCP behavior in this CRD. Therefore I think we should just say:
1. MTU is optional and if it is not provided, OVNK will use the globally configured value for this network.
2. The default global value is 1400.
3. I think we should refrain from listing min/max values for now and not add more work for ourselves.

Reply:
Thanks for explaining :slightly_smiling_face: 

Are we good on the current state of the MTU description?
Or it should also say "MTU is optional, if not provided the global default is used for the network"?

Asking ot make sure we are all on the same page

Reply:
I think its a good idea to be explicit and say:
"MTU is optional, if not provided the globally configured value in OVN-Kubernetes (defaults to 1400) is used for the network"

Reply:
Done